src/main/java/com/amazon/opendistroforelasticsearch/jobscheduler/scheduler/JobScheduler.java

```diff
@@ -164,7 +166,11 @@ boolean reschedule(ScheduledJobParameter jobParameter, JobSchedulingInfo jobInfo
             }
         }
 
-        jobInfo.setExpectedExecutionTime(nextExecutionTime);
+        // Add the defined delay to the job execution
+        if (jobParameter.getDelaySeconds() != null) {
```

Comment:
Add some limitation for `delaySeconds`? The delay should not be negative, and seems not reasonable to delay months.

Reply:
Agree on not allowing a negative delay.. in terms of setting an arbitrary maximum I think we could leave that up to the plugins that utilize this property instead as it might be use case dependent?

Some thoughts - should we allow a plugin to delay a job **past** the next execution? Let's say I have an Alerting Monitor and real time isn't super important, but being correct is. I might set up a monitor for hourly but then a delay of 5 hours because I know data might come a couple hours late. This would be larger than the interval itself.. but is that bad? When the monitor runs 5 hours late it still has the execution windows of the original interval. Is there a different way to write the monitor instead?

Reply:
Make sense, it's necessary to limit the time for delay. But it's hard to determine the upper limit.

Reply:
@ftianli-amzn It seems like there will be an issue if it does go over the next execution time in terms of the scheduling btw.

Scenario: You have a monitor that runs every 1 hour
Executions:  [1PM, 2PM]
Ctx times: [[12, 1], [1, 2]]

You add a delay of 30 minutes
Executions:  [1:30PM, 2:30PM]
Ctx times: [[12, 1], [1, 2]]

When the job is being scheduled for 1:00PM it adds a delay of 30min and then runs at 1:30PM
When that job executes it sees the next execution should be 2PM (i.e. in 30 minutes) and adds a 30min delay on top of that and runs at 2:30PM.

You add a delay of 90 minutes, you would want it to be:
Executions:  [2:30PM, 3:30PM]
Ctx times: [[12, 1], [1, 2]]

But, in reality.. the job will schedule itself for 1pm + 90min delay so it will run at 2:30PM, once it runs it calculates the next time it should run is at 3:00PM, and then adds 90min on that which is now 4:30PM and it completely missed the [1pm ,2pm] interval.

This might be further complicated in cron jobs.

With all that being said we might want to put a constraint of the delay cannot exceed the next execution time and if it does we just take the minimum of the two.

What do you think?

Reply:
I understand in this example, missing a job execution is unacceptable. Although in some case missing one execution may not be a problem, it's necessary to try to prevent complicated situation which may cause error. Taking the minimum of the delay duration and the interval to next execution is a pretty good idea.

Reply:
I'm concerning if setting the delay limit to the next execution interval works for all the cases. I think a similar topic was discussed when adding the `jitter` parameter, long GC pauses and long job executions may lead to a later execution is skipped.

Reply:
Updated the limitation of `delay` in `reschedule()`, and added unit test to test shorten the delay when it exceeds the next execution time.